### PR TITLE
Be able to set the input parameter type for custom node

### PR DESCRIPTION
**Pull request by ke-yu**
This pull request is to fix defect [MAGN 3740 Should be able to set the input parameter type for custom node](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-3740).

Following summarizes what have done in this pull request:
  1. For ``Input`` node, we replace its text control ``DynamoTextBox`` with ``ParameterEditorControl`` which support auto-completion (thanks @aparajit-pratap):

``Input`` node accepts four kinds of input formats:
* ``parameter_name``
* ``parameter_name : type``
* ``parameter_name = default_value``
* ``parameter_name : type = default_value``
  2. It provides limited error handling

"Limited error handling" means we just check the syntax but not semantics. For example, it won't give warning for input ``x : bool = 42``. 
  3. Support default value

Four types are supported
* ``int``
* ``double``
* ``boolean``
* ``string``
  4. Support lacing

  5. Infer input type automatically from ``DSFunction`` node and from ``Function`` node:

Some limitations and issues:
  1. For old custom nodes, their lacing strategy is always ``Disabled``. Shall we enable lacing for them? 
  2. Auto-completion control for ``Input`` node has some bugs, for example, when typing, no item is selected until pressing up/down arrow key. I'm trying to fix it...
  3. If custom node doesn't have any input, then lacing strategy should be disabled. 
  4. As ``Input`` node now has format requirement, it may break old file. For example, a test case tries to verify the nick name of a custom node input port is ``y = f(x)`` (this string is from ``Input``'s ``InputSymbol``). But now we may give a warning. Do we need a migration strategy?

Reviewers
- [x] @ikeough 
- [x] @Benglin
- [ ] @lukechurch  

@riteshchandawar , @monikaprabhu , you may be interested in what have been done for custom node. 

Thanks.

### CODE REVIEW

**Comment by ke-yu** [comment]
Reviewers, I've re-open this pull request after fixing merge conflicts and regressions. Please help to review it :-)

File: src/DynamoCore/DSEngine/CodeCompletionServices.cs [hunk 1]
```diff
@@ -127,6 +127,37 @@ internal IEnumerable<CompletionData> SearchCompletions(string stringToComplete,
         }
```
Comment (Benglin) [inline src/DynamoCore/DSEngine/CodeCompletionServices.cs]: While I was looking to see if this method can be merged with `SearchCompletions` above, I noticed that `Guid guid` parameter is not used, perhaps @aparajit-pratap can remove it separately.

File: src/DynamoCore/DynamoCore.csproj [hunk 2]
```diff
@@ -329,4 +329,4 @@ xcopy /Y "$(SolutionDir)..\extern\SimplexNoise\*" "$(OutputPath)"
   <Target Name="AfterBuild">
```
Comment (Benglin) [inline src/DynamoCore/DynamoCore.csproj]: Might want to revert this file.

File: src/DynamoCoreWpf/Views/Input/ParameterEditor.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void OnTextAreaLostFocus(object sender, RoutedEventArgs e)
+        {
+            InnerTextEditor.TextArea.ClearSelection();
+
+            nodeViewModel.DynamoViewModel.ExecuteCommand(
+                new DynCmd.UpdateModelValueCommand(
+                    nodeViewModel.NodeModel.GUID, "InputSymbol",
+                    InnerTextEditor.Text));
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Input/ParameterEditor.xaml.cs]: Please do not send the command if the text contents are identical (e.g. losing focus without actual changes). That has caused issues with undo [which was subsequently fixed](https://github.com/DynamoDS/Dynamo/pull/3532/files#diff-05003f9e6533825cf76ac9a0a86e5109R432), do look at the changes for details.

**Comment by Benglin** [comment]
Hi @ke-yu, after you have fixed that `UpdateModelValueCommand`, this LGTM.

File: src/DynamoCoreWpf/Views/Input/ParameterEditor.xaml.cs (hunk 3 above)
Comment (ke-yu) [inline src/DynamoCoreWpf/Views/Input/ParameterEditor.xaml.cs]: Thanks for pointing out that. Updated.

**Comment by ke-yu** [comment]
Hi @ikeough , @lukechurch , could you please review the changes?

**Comment by ikeough** [comment]
@ke-yu Psyched to see this in master. LGTM.

**Pull request merged**